common/src/lru.rs

```diff
@@ -1,925 +1,286 @@
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// use mc_common::LruCache;
-    /// let mut cache = LruCache::new(2);
-    ///
-    /// cache.put(1, "a");
-    /// cache.put(2, "b");
-    ///
-    /// assert_eq!(cache.peek(&1), Some(&"a"));
-    /// assert_eq!(cache.peek(&2), Some(&"b"));
-    /// ```
-    pub fn peek<'a, Q>(&'a self, k: &Q) -> Option<&'a V>
-    where
-        KeyRef<K>: Borrow<Q>,
-        Q: Hash + Eq + ?Sized,
-    {
-        match self.map.get(k) {
-            None => None,
-            Some(node) => Some(&node.val),
-        }
-    }
+            // Grab the old entry so that we could return the old value.
+            let prev_entry = self.entries[entry_idx].take();
 
-    /// Returns a mutable reference to the value corresponding to the key in the cache or `None`
-    /// if it is not present in the cache. Unlike `get_mut`, `peek_mut` does not update the LRU
-    /// list so the key's position will be unchanged.
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// use mc_common::LruCache;
-    /// let mut cache = LruCache::new(2);
-    ///
-    /// cache.put(1, "a");
-    /// cache.put(2, "b");
-    ///
-    /// assert_eq!(cache.peek_mut(&1), Some(&mut "a"));
-    /// assert_eq!(cache.peek_mut(&2), Some(&mut "b"));
-    /// ```
-    pub fn peek_mut<'a, Q>(&'a mut self, k: &Q) -> Option<&'a mut V>
-    where
-        KeyRef<K>: Borrow<Q>,
-        Q: Hash + Eq + ?Sized,
-    {
-        match self.map.get_mut(k) {
-            None => None,
-            Some(node) => Some(&mut node.val),
-        }
-    }
+            // Store the new entry in place of the old one.
+            self.entries[entry_idx] = Some((key.clone(), val));
+            self.key_to_entry_index.insert(key, entry_idx);
 
-    /// Returns the value corresponding to the least recently used item or `None` if the
-    /// cache is empty. Like `peek`, `peek_lru` does not update the LRU list so the item's
-    /// position will be unchanged.
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// use mc_common::LruCache;
-    /// let mut cache = LruCache::new(2);
-    ///
-    /// cache.put(1, "a");
-    /// cache.put(2, "b");
-    ///
-    /// assert_eq!(cache.peek_lru(), Some((&1, &"a")));
-    /// ```
-    pub fn peek_lru<'a>(&'a self) -> Option<(&'a K, &'a V)> {
-        if self.len() == 0 {
-            return None;
-        }
+            // Move the entry to the front of the used list.
+            self.used_indexes.remove(used_index);
+            self.used_indexes.push_front(entry_idx);
 
-        let (key, val);
-        unsafe {
-            let node = (*self.tail).prev;
-            key = &(*node).key;
-            val = &(*node).val;
+            return prev_entry.map(|(_k, v)| v);
         }
 
-        Some((key, val))
-    }
+        // Entry not present in cache, see if we have a free slot for it.
+        if let Some(free_index) = self.free_indexes.pop_back() {
+            assert!(self.entries[free_index].is_none());
```

Comment:
Maybe another case where this method should return a `Result`, also wrt the expect and assert below as well.

Reply:
I don't think so. This asset failing means there is a bug in the code - it should never happen.